###### Content under Creative Commons Attribution license CC-BY 4.0, code under BSD 3-Clause License © 2020 Adam Wickenheiser

# Maximizing Thermodynamic Efficiency

This module, consisting of 4 lessons, introduces the idea of *quality* to our discussion of energy we developed in the previous module.  We first think of energy as a quantity: how many joules of energy does it take to move a car a certain distance or to heat up a certain amount of food, for example.  It turns out that this is not a complete picture.  The air in your room contains a large amount of energy in the form of the kinetic and vibrational energy of the molecules making up the air.  However, it is not really practical to try to harness this energy to do something productive.  Another comparison is a small amount of flowing water vs. a large amount of still water.  They may store the same amount of total kinetic energy, but the first case is more useful for hydroelectric energy conversion.

In this lesson, we introduce the idea of *efficiency*, which we use as a comparitive: how well does our proposed design match up with the theoretical "best" design for a given scenario.  To do this, we will have to determine a way to calculate the "best" design, even if it can't actually be built.  This still has practical value though; it gives you a "gold standard" to compare any practical design against, and hence a means of comparing any two practical designs against each other.

## Learning objectives:

1. To be able to describe the theoretical and practical barriers to achieving 100% conversion of heat into useful work.
1. To be able to distinguish between reversible and irreversible processes and what difference this makes in the efficiecy of a process.
1. To be able to describe why both a hot and a cold reservoir are required to produce work from heat or to heat/cool and interior space.
1. To be able to calculate the efficiency of common thermodynamic cycles such as power generators, heat pumps, and refrigerators.
1. To be able to describe the Carnot Cycle as an ideal thermodynamic cycle with maximum efficiency.

## Contents:

1. [How much energy does it take to heat a house?](#heathouse)
1. [Reversible and Irreversible Processes](#rev)
1. [Efficiency and Maximum Efficiency](#eff)
1. [The Carnot Cycle: A Model for Maximum Efficiency](#carnot)

<a id='heathouse'></a>

### How much energy does it take to heat a house?

Let's start with the question posed above: How much energy does it take to heat a house?  If a house were perfectly insulated, no heat would escape, and so theoretically it wouldn't take any energy to maintain a constant indoor temperature.  Of course, a house isn't perfectly insulated, and so some heat escapes based on the materials and thicknesses of the walls and the temperature difference between inside and outside.  This is the essense of design for heat conduction, one of the fundamental processes in the study of Heat Transfer.  In this lesson, we will assume the rate of heat loss is known and try to design a heating system to combat it.

Let's define a few variables:

$T_L$ - the low temperature reservoir, in this case the outside temperature
$T_H$ - the high temperature reservoir, in this case the inside of the house
$\dot{Q}_{loss}$ - the heat loss rate through the walls of the house

By **reservoir** we mean a large body of fluid (usually air or water) at a fixed temperature from which heat can be transferred without changing its temperature.  We assume each of these 3 variables are fixed, known values.

<img src="../images/house heating.png" width="500">

Let's assume the following values: $T_L = 10 ^\circ C$, $T_H = 25 ^\circ C$, and $\dot{Q}_{loss} = 17 kW$.  So, if the house is losing heat at a rate of $17 kW$, how much power is required to keep the inside temperature constant?  It makes sense that it would require $17 kW$ of power to do this, in light of the following diagram:

<img src="../images/house heating energy balance.png" width="300">

It stands to reason that the if the energy in equals the energy out, then the enegy within the system should remain constant, which is absolutely true.  However, it is *not* true that all of the energy input has to come from electrical energy.

Let's look at a device called a **heat pump**.  The following diagram shows the components that make up the heat pump and its thermodynamic cycle.

<img src="../images/heat pump.png" width="500">

The heat pump works exactly the same way as a refrigerator: it takes energy from a cold reservoir and expels it into a warm reservoir.  Thus, both a heat pump and a refrigerator are designed to combat the natural tendency of warm spaces to cool off and cold spaces to warm up, so that everything reaches a thermal equilibrium.  The difference is a refrigerator is designed to maintain the cold space at a specific temperature regardless of the surrounding (wamer) temperature, whereas a heat pump is designed to maintain the warm space at a specific temperature regardless of the surrounding (colder) temperature.

Let's analyze a specific heat pump designed for the house heating situation described above.

**Example:** A heat pump is designed to heat a house, providing heat at a rate of $\dot{Q}_H = 17 kW$.  It uses R-134a as its working fluid, the evaporator operates at a pressure of $P_1 = P_4 = 345 kPa$, and the evaporator operates at a pressure of $P_2 = P_3 = 830 kPa$.  The refrigerant leaves the evaporator as a saturated vapor, leaves the condenser as a saturated liquid, and enters the condenser at $T_2 = 40 ^\circ C$.  Find the power required to run the pump, and the rate of heat transfer from the cold reservoir.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import cantera as ct

In [2]:
QHdot = -17.0e3

# state 1
state1 = ct.Hfc134a()  # create Hfc134a object
P1 = 345.0e3            # pressure [Pa]
x1 = 1.0               # quality
state1.PX = P1, x1     # set pressure and quality
h1 = state1.h          # specific enthalpy [J/kg]
T1 = state1.T

# state 2
state2 = ct.Hfc134a()  # create Hfc134a object
P2 = 830.0e3            # pressure [Pa]
T2 = 40 + 273.15       # temperature [K]
state2.TP = T2, P2     # set temperature and pressure
h2 = state2.h          # specific enthalpy [J/kg]

# state 3
state3 = ct.Hfc134a()  # create Hfc134a object
P3 = P2                # pressure [Pa]
x3 = 0.0               # quality
state3.PX = P3, x3     # set pressure and quality
h3 = state3.h          # specific enthalpy [J/kg]
T3 = state3.T

# state 4
state4 = ct.Hfc134a()  # create Hfc134a object
P4 = P1                # pressure [Pa]
h4 = h3                # temperature [K]
state4.HP = h4, P4     # set specific enthalpy and pressure
h4 = state4.h          # specific enthalpy [J/kg]

mdot = QHdot/(h3-h2)        # mass flow rate [kg/s]
Wdot = mdot*(h1-h2)         # compressor power [W]
Qdot_cond = mdot*(h3-h2)    # condenser heat transfer rate [W]
Qdot_evap = mdot*(h1-h4)    # evaporator heat transfer rate [W]

print('Compressor power =',Wdot/1000.0,'kW')
print('Condenser heat transfer rate =',Qdot_cond/1000.0,'kW')
print('Evaporator heat transfer rate =',Qdot_evap/1000.0,'kW')
print('Condenser temperature =',T3 - 273.15,'C')
print('Evaporator temperature =',T1 - 273.15,'C')

Compressor power = -2.159003245359283 kW
Condenser heat transfer rate = -17.0 kW
Evaporator heat transfer rate = 14.840996754640717 kW
Condenser temperature = 32.62661489646297 C
Evaporator temperature = 4.614814717501247 C


These results show that the pump only requires $2.16 kW$ of power to heat the house.  The rest of the power is being pulled from the air outside into the evaporator at a rate of $14.84 kW$.  So, the flow of energy into the house using a heat pump looks like this:

<img src="../images/house heating with heat pump.png" width="500">

First, we can see that, indeed, the house must receive heat at a rate of $17 kW$ to balance out the heat loss in order to maintain a constant indoor temperature.  As discussed in the previous lesson, **Conservation of energy** tells us that

$\dot{E}_{in} - \dot{E}_{out} = \dot{E}_{CV}$

where $\dot{E}_{in}$ is the rate of energy entering the control volume, $\dot{E}_{out}$ is the rate of energy leaving the control volume, and $\dot{E}_{CV}$ is the rate of change of energy in the control volume.

If we make the steady-state assumption, we get $\dot{E}_{CV} = 0$, and so we can write

$\dot{E}_{in} = \dot{E}_{out}$

If we consider the **control volume** for the house, this means that $\dot{E}_{in} = 17 kW$ must equal $\dot{E}_{out} = 17 kW$.

For the heat pump, this means that $\dot{E}_{in} = 14.84 + 2.16 kW$ must equal $\dot{E}_{out} = 17 kW$.

So no where is the principle of conservation of energy being violated.  The advantage of the heat pump is that the majority of the energy needed to heat the house is drawn from the outdoors (which is free energy!) rather than from electricity.

How do we pull this off?  We take advantage of the fact that the refrigerant is naturally very cold at atmospheric pressures or even a little above.  In this example, the evaporator is operating at $P_1 = 345 kPa$ and $T_4 = 4 ^\circ C$, which is colder than the outside temperature.  Thus, heat flows from the outside into the evaporator, which evaporates the refrigerant from a liquid-vapor mixture to a saturated vapor.  Pumping the refrigerant raises its temperature to $T_2 = 40 ^\circ C$ without much work (only $2.16 kW$), which is hotter than the desired indoor temperature.  Heat is then lost to the indoors through the condenser.  The expansion valve reduces the refrigerant back to $T_4 = 4 ^\circ C$ without requiring any work.  Then the cycle repeats.

<a id='rev'></a>

### Reversible and Irreversible Processes

You may be wondering at this point if we can design an even more efficient heat pump for this house.  After all, the pressures and temperatures of the condenser and evaporator were just given; there's no reason why different values couldn't be chosen.  The evaporator temperature must be lower than the outdoor temperature for heat to flow into the evaporator; similarly, the condenser temperature must be hotter than the indoor temperature for heat to flow into the house.  Other than that, we are free to play around with the values.  Even if you keep all of the pressures and temperatures the same, you might wonder if there is a more efficient pump that can pump the refrigerant from $P_1 = 345 kPa$ up to $P_2 = 800 kPa$.  This lesson and the next will give you the analysis tools to answer these questions.

As a first thought, you may wonder if the required pump power can be made arbitrarily small.  For example, what if it were 0?  Then we would have the following situation:

<img src="../images/house heating no work.png" width="400">

This would mean that heat flows into and out of the house at the same rate, without any external work or coersion, maintaining the temperature of the indoors constant.  However, we know this situation is impossible, just like a hot cup of coffee does not stay hot forever in a cooler room by exchanging equal amounts of energy with it.  Naturally, energy flows from the hot substance to the colder one, not the other way around.  The cup of coffee will naturally cool until it matches the temperature of the room, heating the room slightly, but it will not spontaneously heat itself while cooling the room slightly.

Why does heat only flow in one direction?  Let's look at the particle simulation of a hot fluid (top) coming into contact with a cold fluid (bottom):

<img src="../images/gas mixing.gif" width="300">

Now, what happens if we reverse the video:

<img src="../images/gas mixing reversed.gif" width="300">

The reversed video does not seem odd until the very end, when the green (initially faster moving) particles all end up in the upper half.  Physical laws like Conservation of Mass, Energy, Momentum, etc. work both forwards and backwards in time, so neither animation violates basic principles.  However, the reversed-time scenario is *statistically unlikely* to occur naturally.  In fact, it so improbable that it is unlikely to happen even once in the entire lifespan of the universe!

This is an example of an **irreversible process**: one that, yes, is physically possible, but is so unlikely to occur that it may as well be impossible.  Thus, at least *some* energy is required to run the heat pump needed to heat the house in our previous example.  This concept is summarized by the following statement:

**Clausius Statement:** A cyclic device cannot tranfer heat from low to high temperature bodies without work input from the environment

Since the heat pump is a cyclic device (the working fluid R-134a goes through a complete thermodynamic cycle, returning to its original state) some external work is required to make it operate, in this case supplied by electricity.

Here is another example of a thermodynamic cycle we've seen before:

<img src="../images/ideal gas turbine ex.png" width="600">

Here, heat from e.g. buring coal or nuclear fission evaporates water, which passed through a turbine to produce electricity.  However, the steam must be cooled back down into a liquid to be pumped up to high pressure again; this cooling releases heat into the environment in the condenser.  An easy and effective way to cool something is to run cold water over it; this is why most power plants are built next to rivers, where cold running water is freely available.  The following statement clarifies this requirement:

**Kelvin-Planck Statement:** A cyclic device cannot receive heat from a single reservoir and produce work

Thus, a power plant cannot take heat from combustion/fission and produce work (electricity) without releasing some heat to the environment, for example by releasing it into a river or the atmosphere.  A power plant as just described is an example of a **heat engine** a system that takes heat from a high temperature source, produces useful work (mechanical or electrical), then (necessarily) rejects some waste heat into the environment.

In fact, the Clausius and Kelvin-Planck Statements are equivalent, as shown below:

<img src="../images/second law violations.png" width="600">

On the left, assume there is a way to produce work from a heat source without having to exhaust any heat to the environment (for example, a power plant that does not use river water for cooling).  This is a violation of the Kelvin-Planck Statement.  Then the work produced must be equal to the heat absorbed, so $W = Q_H$.  Let's say this work is used to power a heat pump, which draws heat from the cold environment and uses it to heat the warm environment, like the house above.  Then the total heat transferred to the warm environment must be $Q_H+Q_H$, the total energy input to the heat pump.

The total heat being drawn from the cold environment is $Q_L$, and the total heat transferred to the cold environment is $-Q_H + (Q_H+Q_L) = Q_L$ (note the direction of the arrows).  This is equivalent to the right-hand side, where a heat pump is used to pump heat from a cold reservoir to a warm one without requiring any work!  This is a violation of the Clausius Statement.

We can also run this argument backwards: starting with the violation of the Clausius Statement on the right-hand side, we can draw up a violation of the Kelvin-Planck Statement on the left.  Therefore, both statements are equivalent!  These statements are known as the **Second Law of Thermodynamics**.

<a id='eff'></a>

### Efficiency and Maximum Efficiency

How do we compare two different heat pumps or heat engines operating between two given temperature reservoirs?  The one we analyzed for heating the house seems pretty efficient: it only requires $2.16 kW$ of electricity to provide $17 kW$ of heat to the house.

Let's define the **coefficient of performance** for the heat pump as follows:

$COP_{HP} = \dfrac{\text{desired output}}{\text{required input}} = \dfrac{Q_H}{W_{net,in}}$

where $Q_H$ is the heat transferred to the high temperature reservoir (the inside of the house, in this case), and $W{net,in}$ is the net work (electrical, in this case) input to the cycle.

From conservation of energy, we get $Q - W = \Delta E$, but for a cycle $\Delta E = 0$ (the working fluid returns to its original state in a cycle, by definition).  Recall the convention that $Q$ is the heat transferred *to* the system, while $W$ is the work done *by* the system.  Thus, $W_{net,in} = -W = -Q = -(Q_L - Q_H) = Q_H - Q_L$.  Therefore,

$COP_{HP} = \dfrac{Q_H}{Q_H - Q_L} = \dfrac{1}{1 - Q_L/Q_H}$

If the device is operating in steady-state, we can write this ratio as

$COP_{HP} = \dfrac{\dot{Q}_H}{\dot{W}_{net,in}} = \dfrac{1}{1 - \dot{Q}_L/\dot{Q}_H}$

For our example, we have

$COP_{HP} = \dfrac{\dot{Q}_H}{\dot{W}_{net,in}} = \dfrac{17 kW}{2.16 kW} = 7.87$

From the definition, we can see that a higher $COP$ is better!

Refrigerators act like heat pumps: using work to take heat from a cold reservoir (the inside of the refrigerator) and reject it to a hot reservoir (the space outside the refrigerator).  However, in the case of refrigerators we are more interested in the heat drawn from the cold reservior, so we define the coefficient of performance for refrigerators slightly differently:

$COP_R = \dfrac{\text{desired output}}{\text{required input}} = \dfrac{Q_L}{W_{net,in}}$

where $Q_L$ is the heat transferred from the low temperature reservoir, and $W_{net,in}$ is the net work (electrical in most cases) input to the cycle.  Conservation of energy gives us the same result as with heat pumps: $W_{net,in} = Q_H - Q_L$.  Therefore,

$COP_R = \dfrac{Q_L}{Q_H - Q_L} = \dfrac{1}{Q_H/Q_L - 1}$

If the device is operating in steady-state, we can write this ratio as

$COP_R = \dfrac{\dot{Q}_L}{\dot{W}_{net,in}} = \dfrac{1}{\dot{Q}_H/\dot{Q}_L - 1}$

Heat engines are handled slightly differently since work is the deisred output, and they should use the least amount of heat input (from burning coal or nuclear fission) possible.  With this in mind, let's define the **thermal efficiency** for the heat engine as follows:

$\eta_{th} = \dfrac{\text{desired output}}{\text{required input}} = \dfrac{W_{net,out}}{Q_H}$

From conservation of energy, we get $Q - W = \Delta E$, but for a cycle $\Delta E = 0$.  Recall the convention that $Q$ is the heat transferred *to* the system, while $W$ is the work done *by* the system.  Thus, $W_{net,out} = W = Q = Q_H - Q_L$.  Therefore,

$\eta_{th} = \dfrac{Q_H-Q_L}{Q_H} = 1 - \dfrac{Q_L}{Q_H}$

If the device is operating in steady-state, we can write this as

$\eta_{th} = \dfrac{\dot{W}_{net,out}}{\dot{Q}_H} = 1 - \dfrac{\dot{Q}_L}{\dot{Q}_H}$

These three cyclic devices are summarized in the following figure:

<img src="../images/thermal efficiency.png" width="800">

So, what is the maximum efficiency possible?  We've already stated that 100% efficiency is impossible; a heat engine must reject at least some heat into the environment.  To answer this, we turn to two statements known as **Carnot Principles**:

1. Between two temperature reservoirs, a reversible heat engine is more efficient than an irreversible heat engine.
1. Between two temperature reservoirs, any two reversible heat engines have the same efficiency

What makes a process irreversible?  We already saw one example: heat transfer between hot and cold bodies.  This process naturally occurs in one direction but not the other.  Friction is another example: friction will cause a moving body to come to rest, but a body at rest will not spontaneously start moving.  A good rule of thumb is to imagine a video of the process: is the video plausible playing either forward or in reverse?  As you may have guessed, all processes are irreversible to some extent.

The first Carnot Principle can be <a href="https://en.wikipedia.org/wiki/Carnot%27s_theorem_(thermodynamics)">proven by contradiction</a>: if it were false it would violate the Clausius Statement.  The second statement follows from the first: if $A$ and $B$ are two reversible heat engines, then we must have $\eta_{th,A} \leq \eta_{th,B}$ and $\eta_{th,B} \leq \eta_{th,A}$, and so $\eta_{th,A} = \eta_{th,B}$.

Therefore, the efficiency of a reversible heat engine can only be a function of temperature.  This leads to

$\eta_{th,rev} = 1 - \dfrac{Q_L}{Q_H} = 1 - \dfrac{f(T_L)}{f(T_H)}$

where $f(T)$ is some function.  Since there is no restriction on this function, we can choose $f(T) = T$.  Then

$\eta_{th,rev} = 1 - \dfrac{T_L}{T_H}$

This is the maximum possible efficiency for a heat engine running between reservoirs with these temperatures.  Please note that temperature in this equation *must be in absolute units*, for example Kelvin.

Using this relationship, we can write down the coefficients of performance for a reversible heat pump and refrigerator:

$COP_{HP,rev} = \dfrac{1}{1 - Q_L/Q_H} = \dfrac{1}{1 - T_L/T_H}$

$COP_{R,rev} = \dfrac{1}{Q_H/Q_L - 1} = \dfrac{1}{T_H/T_L - 1}$

In the house heating example, we have $T_L = 10 ^\circ C = 283.15 K$ and $T_H = 25 ^\circ C = 298.15 K$.  Therefore, the maximum $COP$ for this scenario, i.e. the $COP$ for a reversible heat pump, is

$COP_{R,rev} = \dfrac{1}{T_H/T_L - 1} = \dfrac{1}{\frac{298.15 K}{283.15 K} - 1} = 18.88$

Let's compare this to the $COP$ of the heat pump we analyzed: $COP_{HP} = 7.87$.  This tells us that there is room for improvement, but of course attaining the maximum value is impossible.  It would require removing all friction, viscosity, and heat transfer from the system.

Let's write a function to determine if a given heat engine is reversible, irreversible, or impossible (i.e. its efficiency is greater than the maximum possible).  The thermal efficiency of a heat engine is given by

$\eta_{th} = 1 - \dfrac{Q_L}{Q_H}$

and the thermal efficiency of a reversible heat engine is

$\eta_{th,rev} = 1 - \dfrac{T_L}{T_H}$

By the first Carnot Principle, reversible heat engines have the highest efficiency among all heat engines between given temperature reservoirs, so we must have $\eta_{th} \leq \eta_{th,rev}$, which means

$\dfrac{T_L}{T_H} \leq \dfrac{Q_L}{Q_H}$

So, we have the following three possibilities:

1. $\eta_{th} < \eta_{th,rev}$ or $\dfrac{T_L}{T_H} < \dfrac{Q_L}{Q_H}$ (irreversible)
1. $\eta_{th} = \eta_{th,rev}$ or $\dfrac{T_L}{T_H} = \dfrac{Q_L}{Q_H}$ (reversible)
1. $\eta_{th} > \eta_{th,rev}$ or $\dfrac{T_L}{T_H} > \dfrac{Q_L}{Q_H}$ (impossible)

In [3]:
def heat_engine_eta(QL,QH,TL,TH):
    
    eta = 1 - QL/QH         # thermal efficiency
    eta_rev = 1 - TL/TH     # thermal efficiency of a reversible heat pump
    
    if eta < eta_rev:
        print('Heat engine is irreversible')
    elif eta == eta_rev:
        print('Heat engine is reversible')
    else:
        print('Heat engine is impossible')
    
    return eta

In [4]:
# example 1
QL = 500.0        # heat transferred to low-temperature reservior [kJ]
QH = 1000.0       # heat transferred from high-temperature reservior [kJ]
TL = 100.0        # low temperature [K]
TH = 500.0        # high temperature [K]

print('Example 1')
eta = heat_engine_eta(QL,QH,TL,TH)   # thermal efficiency
print('Thermal efficiency =',eta)


# example 2
QL = 600.0        # heat transferred to low-temperature reservior [kJ]
QH = 1000.0       # heat transferred from high-temperature reservior [kJ]
TL = 300.0        # low temperature [K]
TH = 500.0        # high temperature [K]

print('Example 2')
eta = heat_engine_eta(QL,QH,TL,TH)   # thermal efficiency
print('Thermal efficiency =',eta)


# example 3
QL = 600.0        # heat transferred to low-temperature reservior [kJ]
QH = 1000.0       # heat transferred from high-temperature reservior [kJ]
TL = 400.0        # low temperature [K]
TH = 500.0        # high temperature [K]

print('Example 3')
eta = heat_engine_eta(QL,QH,TL,TH)   # thermal efficiency
print('Thermal efficiency =',eta)

Example 1
Heat engine is irreversible
Thermal efficiency = 0.5
Example 2
Heat engine is reversible
Thermal efficiency = 0.4
Example 3
Heat engine is impossible
Thermal efficiency = 0.4


<a id='carnot'></a>

### The Carnot Cycle: A Model for Maximum Efficiency

Can we imagine a heat engine that is completely reversible, and hence has the maximum efficiency among all heat engines between two given temperature reservoirs?  As we previously mentioned, friction and heat transfer between hot and cold bodies are irreversible processes, and so if we can minimize these we can (theoretically) approach a reversible process or cycle.  Although this cannot be attained in practice, it provides an upper bound to the expected thermal efficiency of any given heat engine.

The **Carnot Cycle** is a 4-step heat engine in which each step is theoretically reversible, and so its thermal efficiency is the maximum possible.  Like other heat engines, it takes in heat energy from a high-temperature reservoir, does some useful work, then rejects heat energy to a low-temperature reservoir.

The Carnot Cycle for heat engines is composed of the following four processes, each of which is necessarily reversible to make the whole cycle reversible:

1. **Reversible Isothermal Expansion:** The gas in the cylinder is in thermal equilibrium with the high temperature reservoir at $T_H$; however, the pressure inside the cylinder is greater than the surroundings, and so it expands.  As it expands, the gas cools a small amount, and so, since the reservoir is now warmer than the gas, some heat is transferred into the cylinder.  If the expansion is slow enough, the temperature of the gas remains constant.  This process is theoretically reversible; as the gas is compressed, it heats up a small amount and transfers a small amount of heat to the reservoir, thus maintaining a nearly constant temperature.
1. **Reversible Adiabatic Expansion:** During this phase, the cylinder is insulated, so there can be no heat transfer to the surroundings.  The gas continues to expand, doing work on the surroundings, while cooling from $T_H$ to $T_L$.  This process can be reversed: the gas starts at $T_L$ and is compressed in an insulated cylinder until its temperature reaches $T_H$.
1. **Reversible Isothermal Compression:** The gas in the cylinder is in thermal equilibrium with the low temperature reservoir at $T_L$; however, the pressure inside the cylinder is less than the surroundings, and so it compresses.  As it compresses, the gas warms a small amount, and so, since the reservoir is now colder than the gas, some heat is transferred out of the cylinder.  If the expansion is slow enough, the temperature of the gas remains constant.  Notice the similarity of this process with the first step.
1. **Reversible Adiabatic Compression:** During this phase, the cylinder is insulated, so there can be no heat transfer to the surroundings.  The gas continues to compress, having work done on it by the surroundings, while warming from $T_L$ to $T_H$.  Notice the similarity of this process with the second step.

<img src="../images/carnot cycle.png" width="600">

Let's revisit the Otto cycle we studied previously and see how it compares to a theoretical Carnot cycle between the same two temperature extremes.  Here is the Otto cycle plotted on a P-v diagram, showing the net work output (the area inside the curve), and the steps where heat is added ($Q_H$) and rejected ($Q_L$) to the surroundings.

<img src="../images/otto cycle with work.png" width="500">

We previously derived a formula for the net work output per cycle when the expansion and compression processes are polytropic, i.e. $Pv^n = const.$:

$W_{net,out} = \dfrac{P_2V_2 - P_1V_1}{1-n} + \dfrac{P_4V_4 - P_3V_3}{1-n}$

or, if $n=1$,

$W_{net,out} = P_3V_3 \ln \left(\dfrac{V_4}{V_3} \right) + P_1V_1 \ln \left(\dfrac{V_2}{V_1} \right)$

(Note, if the expansion and compression processes are not polytropic, then the work can be found by numerical integration.)

To compute the thermal efficiency of the Otto cycle, we also need to know the heat input.  Heat is input during the process $2 \to 3$, which models the combustion step in a gasoline engine.  No work is done during this step (because the volume is constant), so conservation of energy gives

$Q_H = m(u_3-u_2)$

For an ideal gas with constant specific heats, we can simplify this as

$Q_H = mc_v(T_3-T_2)$

Let's used these formulas to compute the thermal efficiency of the Otto cycle examined previously, and compare it to a Carnot cycle between the same two temperature extremes.

**Example:** A piston-cylinder operates on the Otto cycle with a compression ratio $r = \frac{v_1}{v_2} = 12$, pressure and temperature at the start of compression $P_1 = 1 atm$ and $T_1 = 18 ^\circ C$, maximum temperature $T_3 = 1300 ^\circ C$, and mass of gas $m = 0.004 kg$.  The expansion and compression steps can be modeled by $Pv^{1.5} = const.$  Compute the thermal efficiency of this cycle, and compare it to a Carnot cycle between the same two temperature extremes.

In [5]:
m = 0.004                     # mass [kg]
T1 = 18.0 + 273.15            # temperature [K]
P1 = 1.0*101325.0             # pressure [Pa]
T3 = 1300.0 + 273.15          # temperature [K]
r = 12.0                      # compression ratio
n = 1.5                       # polytropic process exponent

# state 1
state1 = ct.Solution('air.xml', 'air')      # create air object
state1.TP = T1, P1                          # set temperature and pressure
v1 = state1.v                               # specific volume [m^3/kg]

# state 2
state2 = ct.Solution('air.xml', 'air')      # create air object
v2 = v1/r                                   # specific volume, using compression ratio [m^3/kg]
P2 = P1*(v1/v2)**n                          # pressure, using polytropic process equation [Pa]
state2.DP = 1/v2, P2                        # set density and pressure
u2 = state2.u                               # specific internal energy [J/kg]

# state 3
state3 = ct.Solution('air.xml', 'air')      # create air object
v3 = v2                                     # specific volume [m^3/kg]
state3.TD = T3, 1/v3                        # set temperature and density
P3 = state3.P                               # pressure [Pa]
u3 = state3.u                               # specific internal energy [J/kg]

# state 4
state4 = ct.Solution('air.xml', 'air')      # create air object
v4 = v1                                     # specific volume [m^3/kg]
P4 = P3*(v3/v4)**n                          # pressure, using polytropic process equation [Pa]
state4.DP = 1/v4, P4                        # set density and pressure

# compute work and heat transferred
W_net_out = m*((P2*v2-P1*v1)/(1-n) + (P4*v4-P3*v3)/(1-n))    # net work output [J]
QH = m*(u3-u2)                                               # heat input [J]

# compute thermal efficiencies
eta_th = W_net_out/QH           # thermal efficiency for the Otto cycle
eta_th_rev = 1 - T1/T3          # thermal efficiency for an equivalent Carnot cycle

print('Thermal efficiency for the Otto cycle =',eta_th)
print('Thermal efficiency for an equivalent Carnot cycle =',eta_th_rev)

Thermal efficiency for the Otto cycle = 0.4553917178919074
Thermal efficiency for an equivalent Carnot cycle = 0.8149254680100435
